Added spilling docs (#21718)
ydb-platform/ydb: ydb/docs/en/core/concepts/spilling.md

# Spilling

## Spilling concept

> **Spilling** is a memory management mechanism that temporarily offloads data exceeding available cluster RAM capacity to external storage. Spilling enables execution of user queries that require processing large data volumes exceeding available RAM.

Author: 1. Sure, it should say "exceeding the available node RAM capacity".
2. I intentionally used external storage here because this section is about more general spilling concepts. Even though we are also considering spilling to blob storage or S3, I think it’s worth explicitly noting that currently we spill to the file system. However, I’d prefer to keep external storage in this conceptual part.
3. You are right. I will explicitly note intermediate in this context.

In data processing systems, including {{ ydb-short-name }}, spilling is essential for:

- processing queries with large data volumes when intermediate results don't fit in RAM;
- executing complex analytical operations (aggregations, sorting, table joins) over large datasets;

> - optimizing query performance through intermediate materialization of part of the data in external memory, which in certain scenarios can accelerate overall execution time.

Author: We have a specific optimization called DqReplicate. It allows to avoid reading the save source twice. But it works only with spilling enabled because it may significantly increase memory consumption. So, this is more dq-specific case

## Spilling principles

### Core concepts

Spilling operates based on the memory hierarchy principle:

1. **Random Access Memory (RAM)** — fast but limited resource.
2. **External storage** — slower but more capacious.

When memory usage approaches the limit, the system:

- serializes part of the data;
- saves it to external storage;
- frees the corresponding memory;
- when necessary — loads data back into memory to continue computations.

## Spilling in {{ ydb-short-name }}

### Spilling architecture in {{ ydb-short-name }}

> The spilling mechanism in {{ ydb-short-name }} includes two main levels: backend - a service for storing data in external storage, and frontend - components that interact with this service and manage memory. Currently disk is used as external storage.

Author: There are no such restrictions. Do you have any information on this?

4. Metadata about data location is kept in memory
5. When necessary, data is loaded back and processed

```sql
-- Example query that may trigger spilling in compute cores
SELECT
    user_id,
    COUNT(*) as orders_count,
    SUM(amount) as total_amount
FROM orders
GROUP BY user_id
ORDER BY total_amount DESC;
```

#### 2. Channel Spilling

This type of spilling operates at the level of data transfer between different query execution stages. Data transfer channels automatically buffer and offload data when buffers overflow. This helps avoid blocking the data-generating node execution even when one of the receiving nodes is not ready to accept data.

**Operation mechanism:**

Data transfer channels continuously monitor their state:

1. **Buffering**: Incoming data accumulates in channel internal buffers
2. **Fill control**: System tracks buffer fill levels
3. **Automatic spilling**: When limits are reached, data is automatically serialized and transferred to Spilling Service
4. **Continued operation**: Channel continues accepting new data after freeing memory space
5. **Recovery**: When the next stage is ready, data is read from external storage and passed further

## See also

> - [Spilling configuration](../devops/configuration-management/configuration-v2/spilling-config.md)

Author: moved

- [{{ ydb-short-name }} monitoring](../devops/observability/monitoring.md)

> - [Performance diagnostics](../troubleshooting/performance/index.md)

Author: Currently, there are no public metrics. I will update these articles as soon as we decide which metrics to make public and publish them.

# Spilling

## Spilling in General

> **Spilling** is a memory management mechanism that temporarily offloads intermediate data arising from computations and exceeding available node RAM capacity to external storage. In {{ ydb-short-name }}, disk storage is currently used for spilling. Spilling enables execution of user queries that require processing large data volumes exceeding available node memory.

Author: While the glossary provides only a definition, this section offers both the definition and a more in-depth explanation of spilling. The following lines, up to line 26, expand on the concept in greater detail.

In data processing systems, including {{ ydb-short-name }}, spilling is essential for:

- processing queries with large data volumes when intermediate results don't fit in RAM;
- executing complex analytical operations (aggregations, table joins) over large datasets;
- optimizing query performance through intermediate materialization of part of the data in external memory, which in certain scenarios can accelerate overall execution time.

Spilling operates based on the memory hierarchy principle:

1. **Random Access Memory (RAM)** — fast but limited.
2. **External storage** — slower but more capacious.

When memory usage approaches the limit, the system:

- serializes part of the data
- saves it to external storage
- frees the corresponding memory
- continues processing the query using data remaining in memory
- loads data back into memory, when necessary, to continue computations

## Spilling in {{ ydb-short-name }}

> {{ ydb-short-name }} implements the spilling mechanism through **Spilling Service** — an [actor service](glossary.md#actor-service) that provides temporary storage for data blobs. Detailed technical information about Spilling Service is available in the [Spilling Service](../contributor/spilling-service.md) section.

Author: Good point! Fixed that

### Types of Spilling in {{ ydb-short-name }}

> {{ ydb-short-name }} implements two main types of spilling operating at different levels of the computational process. These types work independently and can activate simultaneously within a single query, providing comprehensive memory management.

Reviewer: ```suggestion
{{ ydb-short-name }} implements two primary types of spilling that operate at different levels of the computational process:

* [Compute Node Spilling](#compute-node-spilling)
* [Channel Spilling](#channel-spilling)

These types work independently and can activate simultaneously within a single query, providing comprehensive memory management.
```

> #### Compute Node Spilling

Reviewer: ```suggestion
#### Compute Node Spilling {#compute-node-spilling}
```

> {{ ydb-short-name }} compute cores automatically offload intermediate data to disk when executing operations that require significant memory. This type of spilling is implemented at the level of individual computational operations and activates when memory limits are reached.
> 
> Main usage scenarios:
> 
> * **Aggregations** — when grouping large data volumes, the system offloads intermediate hash tables to disk  
>   
> * **Join operations** — when joining large tables, the Grace Hash Join algorithm is used with data partitioning and offloading to disk  

Reviewer: ```suggestion
{{ ydb-short-name }} compute cores automatically offload intermediate data to disk when executing operations that require significant memory. This type of spilling is implemented at the level of individual computational operations and activates when memory limits are reached.

Main usage scenarios:

* **Aggregations** — when grouping large data volumes, the system offloads intermediate hash tables to disk.
* **Join operations** — when joining large tables, the [Grace Hash Join](https://en.wikipedia.org/wiki/Hash_join#Grace_hash_join) algorithm is used with data partitioning and offloading to disk.
```

> ##### Operation mechanism

Reviewer: ```suggestion
##### Operation Mechanism
```

Compute nodes contain specialized objects for monitoring memory usage. When data volume approaches the set limit:

1. The system switches to spilling mode
2. Data is serialized and divided into blocks (buckets)
3. Part of the blocks is transferred to the Spilling Service for disk storage
4. Metadata about data location is kept in memory
5. The system continues processing data remaining in memory, which allows freeing additional space
6. When necessary, data is loaded back and processed

> #### Channel Spilling

Reviewer: ```suggestion
#### Channel Spilling {#channel-spilling}
```

4. Metadata about the data location is kept in memory.
5. The system continues processing the data remaining in memory, which frees additional space.
6. When necessary, data is loaded back and processed.

#### Channel Spilling {#channel-spilling}

This type of spilling operates at the level of data transfer between different query execution stages. Data transfer channels automatically buffer and offload data when buffers overflow. This helps avoid blocking the execution of the data-generating node, even when one of the receiving nodes is not ready to accept data.

##### Operation Mechanism

Data transfer channels continuously monitor their state:

1. **Buffering**: Incoming data accumulates in the channel’s internal buffers  
2. **Fill control**: The system tracks buffer fill levels  
3. **Automatic spilling**: When limits are reached, data is automatically serialized and transferred to the Spilling Service  
4. **Continued operation**: The channel continues accepting new data after freeing memory space  
5. **Recovery**: When the next stage is ready, data is read from external storage and passed further  

## Interaction with Memory Controller

> When executing queries, {{ ydb-short-name }} tries to stay within the memory limit set by the [memory controller](../../reference/configuration/index.html#memory-controller). To continue fitting within this limit even as intermediate computations grow, spilling is used. For more details, see the [Memory Management section](../reference/configuration/table_service_config.html#memory-management).

Reviewer: ```suggestion
When executing queries, {{ ydb-short-name }} tries to stay within the memory limit set by the [memory controller](../../reference/configuration/index.md#memory-controller). To continue fitting within this limit as intermediate computations grow, spilling is used. For more details, see the [Memory Management section](../reference/configuration/table_service_config.md#memory-management).
```

## See Also

- [Spilling Service](../contributor/spilling-service.md)

> - [Spilling configuration](../reference/configuration/table_service_config.html)

Reviewer: ```suggestion
- [Spilling configuration](../reference/configuration/table_service_config.md)
```
